lib/logdna/client.rb

```diff
@@ -44,95 +50,125 @@ def process_message(msg, opts = {})
       processed_message
     end
 
-    def schedule_flush
-      start_timer = lambda {
-        sleep(@exception_flag ? @retry_timeout : @flush_interval)
-        flush if @flush_scheduled
-      }
-      Thread.new { start_timer.call }
+    def write_to_buffer(msg, opts)
+      processed_message = process_message(msg, opts)
+      message_size = processed_message.to_s.bytesize
+
+      buffer_byte_size = @lock.synchronize do
+        @buffer.push(Message.new(processed_message, message_size))
+        @buffer_byte_size += message_size
+      end
+
+      if buffer_byte_size >= @flush_threshold
+        @flush_scheduler.unschedule_flush
+        flush
+      else
+        @flush_scheduler.schedule_flush
+      end
     end
 
-    def write_to_buffer(msg, opts)
-      if @lock.try_lock
-        processed_message = process_message(msg, opts)
-        new_message_size = processed_message.to_s.bytesize
-        @buffer.push(processed_message)
-        @buffer_byte_size += new_message_size
-        @flush_scheduled = true
-        @lock.unlock
-
-        if @flush_limit <= @buffer_byte_size
-          flush
+    ##
+    # Asynchronously flush all logs in the buffer
+    def flush
+      Concurrent::Future.execute({ executor: @thread_pool }) { flush_sync }
+    end
+
+    ##
+    # Concurrently flush all logs in the buffer
+    def flush_sync
+      slices = @lock.synchronize do
```

Comment:
If the buffer was threadsafe, could this lock be removed?

Reply:
Yeah the lock is only there to control access to `@buffer` and `@buffer_byte_size`